File: typescript-sdk/src/api/properties.ts
/**
 * PropertiesApi — CRUD for cleaning locations, plus cleaner assignment, iCal sync, and checklists.
 */

import { HttpClient } from "../http-client";
import {
  Property,
  CreatePropertyRequest,
  EnableDisablePropertyRequest,
  AssignCleanerToPropertyRequest,
  ICalRequest,
} from "../models/property";
import { ApiResponse } from "../models/response";

export class PropertiesApi {
  constructor(private readonly http: HttpClient) {}

  /**
   * List all properties, optionally filtering by service type.
   * @param serviceId  Filter by service type ID. Omit to return all.
   */
  listProperties(serviceId?: number): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/properties", serviceId !== undefined ? { serviceId } : undefined);
  }

  /**
   * Add a new property.
   * @param request  name, address, city, country, roomCount, bathroomCount, serviceId.
   * @returns ApiResponse with the created Property.
   */
  addProperty(request: CreatePropertyRequest): Promise<ApiResponse<Property>> {
    return this.http.post<Property>("/v1/properties", request);
  }

  /**
   * Get details of a specific property.
   * @param propertyId  The property ID.
   */
  getProperty(propertyId: number): Promise<ApiResponse<Property>> {
    return this.http.get<Property>(`/v1/properties/${propertyId}`);
  }

  /**
   * Update an existing property.
   * @param propertyId  The property ID.
   * @param request     Updated property fields.
   */
  updateProperty(propertyId: number, request: CreatePropertyRequest): Promise<ApiResponse<Property>> {
    return this.http.put<Property>(`/v1/properties/${propertyId}`, request);
  }

  /**
   * Update additional/supplemental information fields for a property.
   * @param propertyId  The property ID.
   * @param data        Key-value pairs of additional fields.
   */
  updateAdditionalInformation(propertyId: number, data: Record<string, unknown>): Promise<ApiResponse<unknown>> {
    return this.http.put(`/v1/properties/${propertyId}/additional-information`, data);
  }

  /**
   * Enable or disable a property.
   * @param propertyId  The property ID.
   * @param request     { enabled: true | false }
   */
  enableOrDisableProperty(propertyId: number, request: EnableDisablePropertyRequest): Promise<ApiResponse<unknown>> {
    return this.http.post(`/v1/properties/${propertyId}/enable-disable`, request);
  }

  /**
   * Permanently delete a property.
   * @param propertyId  The property ID.
   */
  deleteProperty(propertyId: number): Promise<ApiResponse<unknown>> {
    return this.http.delete(`/v1/properties/${propertyId}`);
  }

  /**
   * Get the list of cleaners assigned to a property.
   * @param propertyId  The property ID.
   */
  getPropertyCleaners(propertyId: number): Promise<ApiResponse<unknown>> {
    return this.http.get(`/v1/properties/${propertyId}/cleaners`);
  }

  /**
   * Assign a cleaner to a property's default cleaner pool.
   * @param propertyId  The property ID.
   * @param request     The cleaner ID to assign.
   */
  assignCleanerToProperty(propertyId: number, request: AssignCleanerToPropertyRequest): Promise<ApiResponse<unknown>> {
    return this.http.post(`/v1/properties/${propertyId}/cleaners`, request);
  }

  /**
   * Remove a cleaner from a property's default cleaner pool.
   * @param propertyId  The property ID.
   * @param cleanerId   The cleaner's user ID.
   */
  unassignCleanerFromProperty(propertyId: number, cleanerId: number): Promise<ApiResponse<unknown>> {
    return this.http.delete(`/v1/properties/${propertyId}/cleaners/${cleanerId}`);
  }

  /**
   * Add an iCal calendar link for availability syncing (Airbnb, VRBO, etc.).
   * @param propertyId  The property ID.
   * @param request     The iCal feed URL.
   */
  addICalLink(propertyId: number, request: ICalRequest): Promise<ApiResponse<unknown>> {
    return this.http.put(`/v1/properties/${propertyId}/ical`, request);
  }

  /**
   * Get the current iCal link for a property.
   * @param propertyId  The property ID.
   */
  getICalLink(propertyId: number): Promise<ApiResponse<unknown>> {
    return this.http.get(`/v1/properties/${propertyId}/ical`);
  }

  /**
   * Remove the iCal calendar link from a property.
   * @param propertyId  The property ID.
   * @param request     The iCal feed URL to remove.
   */
  removeICalLink(propertyId: number, request: ICalRequest): Promise<ApiResponse<unknown>> {
    return this.http.delete(`/v1/properties/${propertyId}/ical`, request);
  }

  /**
   * Set a default checklist for a property.
   * @param propertyId               The property ID.
   * @param checklistId              The checklist ID.
   * @param updateUpcomingBookings   If true, applies to all upcoming bookings too.
   */
  assignChecklistToProperty(
    propertyId: number,
    checklistId: number,
    updateUpcomingBookings = false,
  ): Promise<ApiResponse<unknown>> {
    return this.http.put(
      `/v1/properties/${propertyId}/checklist/${checklistId}?updateUpcomingBookings=${updateUpcomingBookings}`,
    );
  }
}

File: python-sdk/cleanster/api/properties.py
"""Properties API — CRUD for cleaning locations, plus cleaners, iCal, and checklists."""

from typing import Any, Dict, Optional

from ..http_client import HttpClient
from ..models.property import Property
from ..models.response import ApiResponse


class PropertiesApi:
    """
    Property management: create, read, update, delete, enable/disable,
    cleaner assignment, iCal calendar sync, and checklist association.
    """

    def __init__(self, http: HttpClient):
        self._http = http

    def list_properties(self, service_id: Optional[int] = None) -> ApiResponse:
        """
        List all properties. Optionally filter by service type.

        Args:
            service_id: Filter by service type ID. None returns all properties.

        Returns:
            ApiResponse with data as a list of property dicts.
        """
        params = {"serviceId": service_id} if service_id is not None else None
        raw = self._http.get("/v1/properties", params=params)
        return ApiResponse.from_dict(raw)

    def add_property(self, request: Dict[str, Any]) -> ApiResponse:
        """
        Add a new property.

        Args:
            request: Dict with keys: name, address, city, country,
                     roomCount, bathroomCount, serviceId.

        Returns:
            ApiResponse with data as a Property object.
        """
        raw = self._http.post("/v1/properties", body=request)
        return ApiResponse.from_dict(raw, data_factory=Property)

    def get_property(self, property_id: int) -> ApiResponse:
        """Get details of a specific property."""
        raw = self._http.get(f"/v1/properties/{property_id}")
        return ApiResponse.from_dict(raw, data_factory=Property)

    def update_property(self, property_id: int, request: Dict[str, Any]) -> ApiResponse:
        """Update an existing property's details."""
        raw = self._http.put(f"/v1/properties/{property_id}", body=request)
        return ApiResponse.from_dict(raw, data_factory=Property)

    def update_additional_information(
        self, property_id: int, data: Dict[str, Any]
    ) -> ApiResponse:
        """Update additional/supplemental information fields for a property."""
        raw = self._http.put(
            f"/v1/properties/{property_id}/additional-information", body=data
        )
        return ApiResponse.from_dict(raw)

    def enable_or_disable_property(
        self, property_id: int, enabled: bool
    ) -> ApiResponse:
        """
        Toggle a property's active/inactive state.

        Args:
            property_id: The property ID.
            enabled:     True to enable, False to disable.

        Returns:
            ApiResponse.
        """
        raw = self._http.post(
            f"/v1/properties/{property_id}/enable-disable",
            body={"enabled": enabled},
        )
        return ApiResponse.from_dict(raw)

    def delete_property(self, property_id: int) -> ApiResponse:
        """Permanently delete a property."""
        raw = self._http.delete(f"/v1/properties/{property_id}")
        return ApiResponse.from_dict(raw)

    def get_property_cleaners(self, property_id: int) -> ApiResponse:
        """Get the list of cleaners assigned to a property."""
        raw = self._http.get(f"/v1/properties/{property_id}/cleaners")
        return ApiResponse.from_dict(raw)

    def assign_cleaner_to_property(
        self, property_id: int, cleaner_id: int
    ) -> ApiResponse:
        """Assign a cleaner to a property's default cleaner pool."""
        raw = self._http.post(
            f"/v1/properties/{property_id}/cleaners",
            body={"cleanerId": cleaner_id},
        )
        return ApiResponse.from_dict(raw)

    def unassign_cleaner_from_property(
        self, property_id: int, cleaner_id: int
    ) -> ApiResponse:
        """Remove a cleaner from a property's default cleaner pool."""
        raw = self._http.delete(
            f"/v1/properties/{property_id}/cleaners/{cleaner_id}"
        )
        return ApiResponse.from_dict(raw)

    def add_ical_link(self, property_id: int, ical_url: str) -> ApiResponse:
        """
        Add an iCal calendar link to a property for availability syncing.

        Args:
            property_id: The property ID.
            ical_url:    The iCal feed URL (e.g. from Airbnb, VRBO).

        Returns:
            ApiResponse.
        """
        raw = self._http.put(
            f"/v1/properties/{property_id}/ical",
            body={"icalLink": ical_url},
        )
        return ApiResponse.from_dict(raw)

    def get_ical_link(self, property_id: int) -> ApiResponse:
        """Retrieve the current iCal link for a property."""
        raw = self._http.get(f"/v1/properties/{property_id}/ical")
        return ApiResponse.from_dict(raw)

    def remove_ical_link(self, property_id: int, ical_url: str) -> ApiResponse:
        """Remove the iCal calendar link from a property."""
        raw = self._http.delete(
            f"/v1/properties/{property_id}/ical",
            body={"icalLink": ical_url},
        )
        return ApiResponse.from_dict(raw)

    def set_default_checklist(
        self,
        property_id: int,
        checklist_id: int,
        update_upcoming_bookings: bool = False,
    ) -> ApiResponse:
        """
        Set a default checklist for a property.

        Args:
            property_id:               The property ID.
            checklist_id:              The checklist ID.
            update_upcoming_bookings:  If True, apply to all upcoming bookings too.

        Returns:
            ApiResponse.
        """
        path = (
            f"/v1/properties/{property_id}/checklist/{checklist_id}"
            f"?updateUpcomingBookings={str(update_upcoming_bookings).lower()}"
        )
        raw = self._http.put(path)
        return ApiResponse.from_dict(raw)
